silence/silence.go

```diff
@@ -91,6 +91,40 @@ func (c matcherIndex) add(s *pb.Silence) (labels.Matchers, error) {
 	return ms, nil
 }
 
+// silenceVersion associates a silence with the Silences version when it was created.
+type silenceVersion struct {
+	id      string
+	version int
+}
+
+// versionIndex is a index into Silences ordered by the version of Silences when the
+// silence was created. The index is always sorted from lowest to highest version.
+//
+// The versionIndex allows clients of Silences.Query to incrementally update local caches
+// of query results. Instead of a new version requiring the client to scan  everything
+// again to get an up-to-date picture of Silences, they can use the versionIndex to figure
+// out which silences have been added since the last version they saw. This means they can
+// just scan the NEW silences, rather than all of them.
+type versionIndex []silenceVersion
+
+// add pushes a new silenceVersionMapping to the back of the silenceVersionIndex. It does not
+// validate the input.
+func (s *versionIndex) add(version int, sil string) {
+	*s = append(*s, silenceVersion{version: version, id: sil})
+}
+
+// find uses a log(n) search to find the first index of the versionIndex which has a
+// version higher than version. If any entries with a higher version exist, it returns true
+// and the starting index (which is guaranteed to be a valid index into the slice). Otherwise
+// it returns false.
+func (s versionIndex) find(version int) (int, bool) {
+	startIdx := sort.Search(len(s), func(i int) bool {
+		return s[i].version > version
+	})
+
+	return startIdx, startIdx < len(s)
+}
```

Comment:
nit: we can use more verbose method naming and return variable naming to make it more clear:
```suggestion
func (s versionIndex) findBiggerVersionIndex(version int) (index int, found bool) {
	startIdx := sort.Search(len(s), func(i int) bool {
		return s[i].version > version
	})

	return startIdx, startIdx < len(s)
}
```